### PR TITLE
Added support for child devices with Outlet and Underbed Light

### PR DESCRIPTION
**Pull request by bsileo**
Here is a PR with some basic support for child devices. In my case, I wanted to include the Underbed Lights in a Scene. I can't do this without a proper switch device in Hubitat, so the child comes in handy. Feel free to not use this in your production code if it does not fit your direction / style etc!

I only did Outlet and Underbed Light for now with basic on/off control. We could certainly add more to this same pattern for other aspects of the bed.

### CODE REVIEW

File: SleepNumberController_Driver.groovy [hunk 1]
```diff
@@ -139,6 +143,88 @@ def updated() {
...
+}
+
+def manageChildren() {
+    switch (state.type) {
+        case "presence":
+            def outlet = getOutlet()
+            if (!outlet) {
+                outlet = addChildDevice('hubitat','Generic Component Switch', getChildDNI("outlet"),
+                                        [ label: getChildName("Outlet"),
+                                         componentName: "outlet",
```
Comment (rvrolyk) [inline SleepNumberController_Driver.groovy:158]: Maybe naive question (but I haven't dealt w/ composite devices before) - the original post on [this](https://community.hubitat.com/t/composite-devices-parent-child-devices/1925) seems to say properties are name and label but you have componentName and componentLabel.  Are these synonymous?

File: SleepNumberController_Driver.groovy [hunk 2]
```diff
@@ -139,6 +143,88 @@ def updated() {
...
+                                         componentLabel: getChildName("Underbed Light"),                	                  
+                                         completedSetup:true])
+                log.info("Created Underbed light child device")
+            }
+            break
+    }
+}
+
+def getOutlet() {
+    return childDevices.find({it.deviceNetworkId == getChildDNI("outlet")})
```
Comment (rvrolyk) [inline SleepNumberController_Driver.groovy:177]: Couldn't you fetch this by name (w/o the iterator) by using getChildDevice(DNI)?

File: SleepNumberController_Driver.groovy [hunk 3]
```diff
@@ -139,6 +143,88 @@ def updated() {
...
+
+def getChildDNI(type, name) {
+    return device.deviceNetworkId + + "-${type}-${name}"
+}
+
+def componentRefresh(device) {
+    log.warn("Component Refresh not implemented")
+}
+
+def componentOn(device) {
```
Comment (rvrolyk) [inline SleepNumberController_Driver.groovy:201]: (another naive question) - is this called when the child device is turned on?

File: SleepNumberController_Driver.groovy (hunk 2 above)
Comment (bsileo) [reply SleepNumberController_Driver.groovy:177]: yes, that probably would be cleaner:

```suggestion
    return getChildDevice(getChildDNI("outlet"))
```

File: SleepNumberController_Driver.groovy (hunk 3 above)
Comment (bsileo) [reply SleepNumberController_Driver.groovy:201]: Yes the generic component devices define a standard API for calls back to the parent.  This is not well documented as far as I know, but you can see the code here:

https://github.com/hubitat/HubitatPublic/blob/master/examples/drivers/componentSwitch.groovy

File: SleepNumberController_Driver.groovy (hunk 1 above)
Comment (bsileo) [reply SleepNumberController_Driver.groovy:158]: You are right this is a carryover from SmartThings code - I took this from another driver I have that was built to be cross-compatible before SmartThingsa dropped the classic app.

So I think we can drop the component* tags.  I'll do some cleanup on this.

**Comment by bsileo** [comment]
One more thing to consider here - this code does not have any support for polling .updates to the child devices at this point. I am not sure if you are getting data about the status of the lights within ProcessBedData()?  If we are getting this status, we can pass it along to these child devices to have them correctly follow the current state such as when it is changed from the bed remote.

**Comment by rvrolyk** [comment]
I hadn't wired outlet status yet merely because the bed light status didn't
seem so interesting but outlet status probably is.  I will need to update
the app to better track availability of the outlets and then wire it up in
#processBedData... My plan is to attempt to review your code, merge into
the beta branch and probably extract a few of the other devices out into
child devices and then I'll see about wiring status transfer from the
parent down.

One random question I'm sure has an obvious answer I'm just missing -- what
is the advantage of a child device over the "shared" device I have now?  Is
it just that you wanted a new switch per outlet and I am not exposing that
at this point?  Just trying to figure out if it's worth moving some of the
current options into child devices... overall it seems cleaner but each
device would be less specific aside from the parent (not sure anyone cares?)

>

**Comment by bsileo** [comment]
The key difference is that this way you can have a device with a given capability for all the various actions. So with the child device, we now have a device with the Switch capability to control the outlet. with the way you had it, these are all custom commands.  While you can access those customcommands from some parts of Hubitat, others you can't.  So in my example I wanted to access this "Switch" from a scene and that App does not support access to custom commands.  In similar fashion, I now exp[ose that switch to Sharptools and it is super easy to make it a button on my dashboard to turn the nightstand lamp (on my outlet) on and off from the dashboard. 

In your code, you are using the type to determine what the "switch" capability does for a single object.  If you create child devices, you could in effect only have a single "Sleep Number" device that represents the whole bed and use child devices for all the features. Or perhaps one device per side with child devices. Of course no right answer here...its the developers' choice what mix and match of custom commands versus child devices makes sense. I sort of like the approach where we have a preference to decide if we create the children because some users may not want all of them. I use this same approach for the Pool Controller device I maintain - In that case, a given Pool Controller might have between 4 and 24 auxiliary circuits which each get a child switch, multiple pumps, temperature sensors etc.  So in that case you really have to use the children since the configuration is dynamic and you can't really build up dynamic commands based on a given user's configuration.

**Comment by rvrolyk** [comment]
That makes sense.  I was attempting to do exactly that by creating new devices with the type indicating what they are and giving users options to install the different ones based on their configuration.  At the time I was unaware of the notion of a child/composite device so didn't take that approach.  So I suppose each of my "types" today could become a child device of the primary "side" device (probably at that level vs. the entire bed since the side is the most useful granularity I suspect).

I'll play with it a bit and see... I'm sure I'll see this but I assume child devices show up nested in the UI and I can drop the container notion that I have now if I do this?

**Comment by bsileo** [comment]
Yes they show up nested:

image

**Comment by rvrolyk** [comment]
Darn.  If I'd known about that I would've written it that way to start ... not sure I'll have time to tinker with it this weekend but will try to make some progress in the next few weeks and see how it goes.  Thanks for the idea!

**Pull request merged**